File: typescript/packages/mpp/src/utils/ed25519.ts
/**
 * Ed25519 precompile instruction builder.
 *
 * Constructs the instruction data for the Ed25519 precompile program
 * (Ed25519SigVerify111111111111111111111111111) to verify a signature
 * over an arbitrary message.
 *
 * The instruction data layout for a single signature (all inline):
 *   [0..2]    num_signatures: u16 LE = 1
 *   [2..4]    signature_offset: u16 LE = 16
 *   [4..6]    signature_instruction_index: u16 LE = 0xFFFF (same instruction)
 *   [6..8]    public_key_offset: u16 LE = 80
 *   [8..10]   public_key_instruction_index: u16 LE = 0xFFFF
 *   [10..12]  message_data_offset: u16 LE = 112
 *   [12..14]  message_data_size: u16 LE
 *   [14..16]  message_instruction_index: u16 LE = 0xFFFF
 *   [16..80]  signature: 64 bytes
 *   [80..112] public_key: 32 bytes
 *   [112..]   message: variable length
 */

import type { Address, Instruction } from '@solana/kit';

const ED25519_PROGRAM_ID = 'Ed25519SigVerify111111111111111111111111111' as Address;

const SAME_INSTRUCTION: number = 0xffff;
// Layout: 2 (num_sigs u16) + 14 (descriptor: 7 x u16) = 16 bytes header
const HEADER_SIZE = 16;
const SIGNATURE_SIZE = 64;
const PUBKEY_SIZE = 32;
const SIGNATURE_OFFSET = HEADER_SIZE;
const PUBKEY_OFFSET = SIGNATURE_OFFSET + SIGNATURE_SIZE;
const MESSAGE_OFFSET = PUBKEY_OFFSET + PUBKEY_SIZE;

/**
 * Build an Ed25519 precompile verify instruction.
 *
 * The instruction tells the precompile to verify that `signature` is a valid
 * Ed25519 signature of `message` by `publicKey`. The runtime verifies the
 * signature; our on-chain program then reads the instructions sysvar to
 * confirm the precompile was asked to verify the correct key and message.
 */
export function createEd25519VerifyInstruction(
    publicKey: Uint8Array,
    signature: Uint8Array,
    message: Uint8Array,
): Instruction {
    if (publicKey.length !== PUBKEY_SIZE) {
        throw new Error(`Public key must be ${PUBKEY_SIZE} bytes, got ${publicKey.length}`);
    }
    if (signature.length !== SIGNATURE_SIZE) {
        throw new Error(`Signature must be ${SIGNATURE_SIZE} bytes, got ${signature.length}`);
    }

    const totalSize = MESSAGE_OFFSET + message.length;
    const data = new Uint8Array(totalSize);
    const view = new DataView(data.buffer);

    // num_signatures
    view.setUint16(0, 1, true);

    // Descriptor (starts at offset 2, immediately after num_signatures)
    view.setUint16(2, SIGNATURE_OFFSET, true); // signature_offset
    view.setUint16(4, SAME_INSTRUCTION, true); // signature_instruction_index
    view.setUint16(6, PUBKEY_OFFSET, true); // public_key_offset
    view.setUint16(8, SAME_INSTRUCTION, true); // public_key_instruction_index
    view.setUint16(10, MESSAGE_OFFSET, true); // message_data_offset
    view.setUint16(12, message.length, true); // message_data_size
    view.setUint16(14, SAME_INSTRUCTION, true); // message_instruction_index

    // Data
    data.set(signature, SIGNATURE_OFFSET);
    data.set(publicKey, PUBKEY_OFFSET);
    data.set(message, MESSAGE_OFFSET);

    return {
        accounts: [],
        data,
        programAddress: ED25519_PROGRAM_ID,
    };
}

File: programs/mpp-channel/src/ed25519.rs
use anchor_lang::prelude::*;

use crate::errors::MppChannelError;

// Ed25519 instruction data layout offsets (for a single signature).
// See: https://docs.solanalabs.com/runtime/programs#ed25519-program
//
// Header:
//   [0..2]   num_signatures (u16 LE) = 1
//   [2..4]   padding (u16 LE) = 0
//
// Per-signature descriptor (16 bytes):
//   [4..6]   signature_offset (u16 LE)
//   [6..8]   signature_instruction_index (u16 LE) = 0xFFFF (same instruction)
//   [8..10]  public_key_offset (u16 LE)
//   [10..12] public_key_instruction_index (u16 LE) = 0xFFFF
//   [12..14] message_data_offset (u16 LE)
//   [14..16] message_data_size (u16 LE)
//   [16..18] message_instruction_index (u16 LE) = 0xFFFF
//
// Padding:
// Data (for inline, all in same instruction):
//   [16..80]  signature (64 bytes)
//   [80..112] public_key (32 bytes)
//   [112..]   message (variable)

const HEADER_SIZE: usize = 2; // num_signatures u16
const DESCRIPTOR_SIZE: usize = 14; // 7 x u16 fields

const SIGNATURE_SIZE: usize = 64;
const PUBKEY_SIZE: usize = 32;

const DATA_START: usize = HEADER_SIZE + DESCRIPTOR_SIZE; // 16
const SIGNATURE_OFFSET: usize = DATA_START;
const PUBKEY_OFFSET: usize = SIGNATURE_OFFSET + SIGNATURE_SIZE;
const MESSAGE_OFFSET: usize = PUBKEY_OFFSET + PUBKEY_SIZE;

/// Validate that a specific instruction in the transaction is an Ed25519
/// precompile verification of the expected public key and message.
///
/// The Ed25519 precompile itself verifies the cryptographic signature.
/// This function verifies that the precompile was asked to check the
/// correct inputs (the payer's public key and the binary voucher bytes).
///
/// If this validation is wrong or missing, anyone could submit a settle/close
/// transaction with an Ed25519 instruction that verifies a different key or
/// message, effectively bypassing signature authorization.
pub fn validate_ed25519_instruction(
    instructions_sysvar: &AccountInfo,
    expected_signer: &Pubkey,
    expected_message: &[u8],
    ed25519_instruction_index: u8,
) -> Result<()> {
    // Load the instruction at the given index from the instructions sysvar.
    let instruction = solana_instructions_sysvar::load_instruction_at_checked(
        ed25519_instruction_index as usize,
        instructions_sysvar,
    )
    .map_err(|_| MppChannelError::MissingEd25519Instruction)?;

    // Verify the instruction targets the Ed25519 precompile program.
    if instruction.program_id != solana_sdk_ids::ed25519_program::ID {
        return Err(MppChannelError::InvalidEd25519Program.into());
    }

    let data = &instruction.data;

    // Minimum size: header + descriptor + padding + signature + pubkey + at least 1 byte message
    let min_size = MESSAGE_OFFSET + 1;
    if data.len() < min_size {
        return Err(MppChannelError::MissingEd25519Instruction.into());
    }

    // Verify num_signatures == 1 (we only support single-signature verification).
    let num_signatures = u16::from_le_bytes([data[0], data[1]]);
    if num_signatures != 1 {
        return Err(MppChannelError::MissingEd25519Instruction.into());
    }

    // Extract the public key from the instruction data and compare.
    let pubkey_bytes = &data[PUBKEY_OFFSET..PUBKEY_OFFSET + PUBKEY_SIZE];
    if pubkey_bytes != expected_signer.as_ref() {
        return Err(MppChannelError::InvalidEd25519PublicKey.into());
    }

    // Extract the message from the instruction data and compare.
    let message_bytes = &data[MESSAGE_OFFSET..];
    if message_bytes != expected_message {
        return Err(MppChannelError::InvalidEd25519Message.into());
    }

    Ok(())
}
